R/create_profile_matrix.R

```diff
@@ -0,0 +1,243 @@
+  #assign cell name to type
+  names(cellTypes) <- cellAnnots[[cellNameCol]]
+  
+  if(is.null(cellTypes)){
+    stop("Individual cell's type are not valid")
+  }
+  
+  rm(cellAnnots)
+  
+  # mtx <- as.data.frame(mtx)
+  
+  if(!any(names(cellTypes) %in% colnames(mtx)) & 
+     any(names(cellTypes) %in% rownames(mtx))){
+    print("Transposing Matrix")
+    mtx<- t(mtx)
+  }
+  
+  if(!any(names(cellTypes) %in% colnames(mtx))){
+    stop(paste("cellNameCol names does not match count matrix column names", 
+               "matrix cell names:", colnames(mtx)[1], "annots cell names:", names(cellTypes)[1]))
+  }else if(!all(names(cellTypes) %in% colnames(mtx))){
+    missing <- length(which(!names(cellTypes) %in% colnames(mtx)))
+    warning(paste("not all cellNameCol names are in count matrix;", missing, "cells are missing"))
+  }
+  
+  if(filterCellTypes == TRUE){
+    #remove cells with no cell type assignment
+    w2rm <- which(is.na(cellTypes) | tolower(cellTypes) %in% c("unspecified", "unknown", "not available")) 
+    w2rm <- unique(c(w2rm, grep(pattern = "doublet|dividing|low q|filtered|mitotic", x = tolower(cellTypes))))
+    if(length(w2rm) > 0){
+      cellTypes <- cellTypes[-w2rm]
+    }
+  }
+  
+  #normalize data if necessary 
+  if(normalize == TRUE){
+    print("Normalizing Matrix")
+    med <- median(Matrix::colSums(mtx))
+    cols <- colnames(mtx)
+    rows <- rownames(mtx)
+    mtx <- Matrix::Matrix(as.matrix(mtx) %*% as.matrix(diag(1/Matrix::colSums(mtx))*med), sparse = T) 
+    
+    colnames(mtx) <- cols
+    rownames(mtx) <- rows
+    
+    rm(cols,rows)
+  }
+  
+  atlas <- NULL
+  
+  #get all unique cell types
+  CTs <- unique(cellTypes)
+  
+  print("Creating Atlas")
+  
+  for(i in CTs){
+    #print log of progress
+    print(paste(which(CTs == i), "/", length(CTs), ":", i))
+    
+    #get cell names for this cell type
+    cellsType <- names(cellTypes)[which(cellTypes == i)]
+    #confirm cells are in matrix
+    cellsType <- cellsType[which(cellsType %in% colnames(mtx))]
+    
+    if(length(cellsType) > minCellNum){
+      
+      if(length(cellsType) > 1){
+        #remove cells with low gene expression
+        cellsType <- cellsType[which(Matrix::colSums(mtx[,cellsType] > 0) > minGenes)]
+      }else{
+        cellsType <- cellsType[which(sum(mtx[,cellsType] > 0) > minGenes)]
+      }
+      
+      
+      if(length(cellsType) > minCellNum){
+        #get average expression if there are enough cells for cell type
+        if(length(cellsType) > minCellNum & length(cellsType) != 1){
+          atlas <- as.data.frame(cbind(atlas, Matrix::rowMeans(mtx[,cellsType], na.rm = T)))
+          colnames(atlas)[ncol(atlas)] <- i
+        }else{
+          atlas <- as.data.frame(cbind(atlas, mtx[,cellsType]))
+          colnames(atlas)[ncol(atlas)] <- i
+        }
+      }
+    }
+  }
+  
+  rm(cellTypes)
+  
+  #subset to genes expressed in at least a user defined number of cell type(s)
```

Comment:
Are you sure we want to do this? I'd think that a gene expressed in just 1 cell type could be a super useful marker gene. Perhaps this check should also make sure the gene isn't somewhat highly expressed in at least 1 cell type? That is, change the rule to keep genes that are 1. expressed in > numCellTypesExpre cell types OR 2. are somewhat high in at least 1 cell type.

Reply:
The default here is an inclusive 1. So by default, we are only removing genes that are 0 across the entire matrix. I also agree that filtering the matrix too far in this step could remove good marker genes. Maybe we remove the user-defined threshold for this filtering step and in a later iteration we can make a more intelligent gene filtering step by CV or something? The number of genes in the final matrix directly contributes to the run time of SpatialDecon so this was a quick look at filtering down the matrix.

Reply:
I like your suggestion above - cut this as an argument for the sake of simplicity / preventing user errors, and consider smarter filtering later.